File: back/src/main/java/com/curso/spring/rest/model/entity/ItemFactura.java
package com.curso.spring.rest.model.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import javax.validation.constraints.Min;
import java.io.Serializable;
import java.math.BigDecimal;

/**
 * Entidad que representa un item de la factura en la base de datos
 */
@Entity
@Table(schema = "tfg", name = "facturas_items")
public class ItemFactura implements Serializable {

    private static final long serialVersionUID = -1142856044917006208L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // relación many to one con producto
    @ManyToOne(fetch = FetchType.LAZY)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    private Producto producto;

    @Min(value = 1, message = "tiene que ser mayor que 0")
    private Integer cantidad;

    private BigDecimal precio;

    public ItemFactura() {}

    public ItemFactura(Integer cantidad, Producto producto) {
        this.setCantidad(cantidad);
        this.setProducto(producto);
        this.setPrecio(producto.getPrecio());
    }

    public Long getId() {
        return this.id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Integer getCantidad() {
        return this.cantidad;
    }

    public void setCantidad(Integer cantidad) {
        this.cantidad = cantidad;
    }

    public BigDecimal getPrecio() {
        return this.precio;
    }

    public void setPrecio(BigDecimal precio) {
        this.precio = precio;
    }

    public Producto getProducto() {
        return this.producto;
    }

    public void setProducto(Producto producto) {
        this.producto = producto;
    }

    /**
     * Calcula el importe
     * @return precio * cantidad
     */
    public BigDecimal getImporte() {
        return this.precio.multiply(BigDecimal.valueOf(this.cantidad));
    }
}

File: front/src/app/models/itemFactura.ts
import {Producto} from './producto';
import {OnInit} from '@angular/core';

export class ItemFactura implements OnInit {

  private _id: number;
  private _producto: Producto;
  private _precio: number;
  private _cantidad: number;
  private _importe: number;

  constructor(producto?: Producto, item?: ItemFactura) {
    if (producto) {
      this.producto = producto;
      this.cantidad = 1;
      this.precio = producto.precio as number;
    } else if (item) {
      this.id = item.id;
      this.producto = item.producto;
      this.cantidad = item.cantidad;
      this.precio = item.precio;
    }
    this.setImporte();
  }

  ngOnInit(): void {
  }

  get id(): number {
    return this._id;
  }

  set id(value: number) {
    this._id = value;
  }

  get producto(): Producto {
    return this._producto;
  }

  set producto(value: Producto) {
    this._producto = value;
  }

  get cantidad(): number {
    return this._cantidad;
  }

  set cantidad(value: number) {
    this._cantidad = value;
  }

  get importe(): number {
    return this._importe;
  }

  set importe(value: number) {
    this._importe = value;
  }

  setImporte() {
    return this._importe = this.precio * this._cantidad;
  }

  get precio(): number {
    return this._precio;
  }

  set precio(value: number) {
    this._precio = value;
  }

  toJSON() {
    return {
      id: this.id,
      producto: this.producto,
      precio: this.precio,
      cantidad: this.cantidad,
      importe: this.importe
    };
  }
}
